gapic-common/lib/gapic/common/retry_policy.rb

```diff
@@ -0,0 +1,187 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "gapic/common/error_codes"
+
+module Gapic
+  module Common
+    # Gapic Common retry policy base class.
+    class RetryPolicy
+      # @return [Numeric] Default initial delay in seconds.
+      DEFAULT_INITIAL_DELAY = 1
+      # @return [Numeric] Default maximum delay in seconds.
+      DEFAULT_MAX_DELAY = 15
+      # @return [Numeric] Default delay scaling factor for subsequent retry attempts.
+      DEFAULT_MULTIPLIER = 1.3
+      # @return [Array<String|Numeric>] Default list of retry codes.
+      DEFAULT_RETRY_CODES = [].freeze
+      # @return [Numeric] Default timeout threshold value in seconds.
+      DEFAULT_TIMEOUT = 3600 # One hour
+
+      ##
+      # Create new Gapic::Common:RetryPolicy instance.
+      #
+      # Instance values are set as `nil` to determine whether values are overriden from default.
+      #
+      # @param initial_delay [Numeric] Initial delay in seconds for this policy.
+      # @param max_delay [Numeric] Maximum delay in seconds for this policy.
+      # @param multiplier [Numeric] The delay scaling factor for each subsequent retry attempt.
+      # @param retry_codes [Array<String|Numeric>] List of retry codes.
+      # @param timeout [Numeric] Timeout threshold value in seconds.
+      #
+      def initialize initial_delay: nil, max_delay: nil, multiplier: nil, retry_codes: nil, timeout: nil
+        @initial_delay = initial_delay
+        @max_delay = max_delay
+        @multiplier = multiplier
+        @retry_codes = convert_codes retry_codes
+        @timeout = timeout
+        @delay = nil
+      end
+
+      def initial_delay
+        @initial_delay || DEFAULT_INITIAL_DELAY
+      end
+
+      def max_delay
+        @max_delay || DEFAULT_MAX_DELAY
+      end
+
+      def multiplier
+        @multiplier || DEFAULT_MULTIPLIER
+      end
+
+      def retry_codes
+        @retry_codes || DEFAULT_RETRY_CODES
+      end
+
+      def timeout
+        @timeout || DEFAULT_TIMEOUT
+      end
+
+      ##
+      # Perform delay if and only if retriable.
+      #
+      # Positional argument `error` takes precedence over positional arguments, meaning the latter
+      # will only be used if and only if `error` is not present.
```

Comment:
This is actually not true. The first positional argument (if present) will be assigned to `error`. Any additional positional arguments will be assigned to `args`.

Reply:
Good catch, I forgot to switch it back. Keeping as `error = nil` and using that parameter to check which retry logic to use.